File: java-client/src/main/java/com/datastr/sort/QuickSort.java
package com.datastr.sort;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

import static java.util.stream.Collectors.toList;

public class QuickSort implements SortAlgo {

    public List<Integer> sort(List<Integer> elements) {
        if (elements.size() == 0) {
            return elements;
        }

        Integer pivot = elements.get(0);


        List<Integer> sortedList = new ArrayList<>();
        sortedList.addAll(sort(filter(elements, integer -> integer < pivot)));
        sortedList.addAll(filter(elements, integer -> integer.equals(pivot)));
        sortedList.addAll(sort(filter(elements, integer -> integer > pivot)));
        return sortedList;
    }

    private List<Integer> filter(List<Integer> elements, Predicate<Integer> integerPredicate) {
        return elements.stream()
                .filter(integerPredicate)
                .collect(toList());
    }

}

File: kotlin-core/src/main/kotlin/com/datastr/sort/QuickSort.kt
package com.datastr.sort

import kotlin.streams.toList

fun quickSort(list: List<Int>): List<Int> {
    return if (list.isEmpty())
        list
    else {
        val sortedList = ArrayList<Int>()
        val pivot = list[0]
        sortedList.addAll(quickSort(filter(list) { it < pivot }))
        sortedList.addAll(filter(list) { it == pivot })
        sortedList.addAll(quickSort(filter(list) { it > pivot }))
        sortedList
    }
}

private fun filter(list: List<Int>, lambda: (t: Int) -> Boolean) =
        list.stream().filter(lambda).toList()